### ISSUE TITLE
Validation_step=None is an error

### ISSUE THREAD

**Issue by Rstln** [tushar-jain01/Image-Colorization#1]
in train.py line

I run your code in pychram and find that Validation_step is not set in model.fit() function while validation data is used, so an error occurs.

By the way, I want to ask how to create the .h5 file which contains weights of the model. I learned that .h5 file is composed of dataset and groups, but I don't know how to organize the model's weights with it, considering the fact that at the beginning no model is saved in the path. If a .h5 file could be uploaded, I'd be appreciated.

**Comment by tushar-jain01** [comment]
Hello!

Thanks for reporting
I have resolved the issue and updated the train.py

## Saving a model
In order to save a model you can use the following Command here you must specify the Model path along with the name of your model and with the extension `.h5` if you want to save a `.h5` file [🔗](https://github.com/tushar-jain01/Image-Colorization/blob/bb667d96af8a1107d13fb5debdacaf1132070d97/Suorce/train.py#L60)``model.save(SAVE_MODEL_PATH)``

I have already uploaded the h5 files you can check for the link in readme.

Thanks

**Comment by Rstln** [comment]
By the way, the paper says the model can handle any resolution but when I load images whose sizes are different, error occurs in function "loadImagesToArray()" when returning. Since np.array requires all the members inside are in the same size, my training data with different size couldn't be loaded. Any solution?

**Comment by Rstln** [comment]
And your advices is helpful! I've learned how to save and load the model/weights!

**Comment by Rstln** [comment]
> 
> 
> By the way, the paper says the model can handle any resolution but when I load images whose sizes are different, error occurs in function "loadImagesToArray()" when returning. Since np.array requires all the members inside are in the same size, my training data with different size couldn't be loaded. Any solution?

--------------------------
”A `Concatenate` layer requires inputs with matching shapes“ also occurs when the height and width of one image are different. I guess the paper input different size images to the *two* low level network. One for mid-level network, and it doesn't require size to be (width==height). The other one for global network, which requires the size to be square. So I suppose set two inputs and two "low level network" is feasible. Looking forward your opinion.

**Comment by tushar-jain01** [comment]
> By the way, the paper says the model can handle any resolution but when I load images whose sizes are different, error occurs in function "loadImagesToArray()" when returning. Since np.array requires all the members inside are in the same size, my training data with different size couldn't be loaded. Any solution?

To load different size images to nd array you can load them in batches with each batch of will contain the images of equal size. in this way you will be able to load different size images.

> > By the way, the paper says the model can handle any resolution but when I load images whose sizes are different, error occurs in function "loadImagesToArray()" when returning. Since np.array requires all the members inside are in the same size, my training data with different size couldn't be loaded. Any solution?
> 
> ”A `Concatenate` layer requires inputs with matching shapes“ also occurs when the height and width of one image are different. I guess the paper input different size images to the _two_ low level network. One for mid-level network, and it doesn't require size to be (width==height). The other one for global network, which requires the size to be square. So I suppose set two inputs and two "low level network" is feasible. Looking forward your opinion.


1. First of you should not specify any input size, you can specify like `Input(shape=(None,None,n_channels))`.
2. Convolutional Layer can handle variable sized input because we deal with kernels but problems come when we use fully connected layer (`Dense` in TensorFlow ).  
3. In order for the model to predict on different sized images you have to make careful use of global average pooling or max pooling. you have to make sure that input to a dense layer should be of same size no matter whatever is your input you can see [Here](https://stackoverflow.com/a/44103746/8172437) and [Here](https://github.com/keras-team/keras/issues/1920#issuecomment-326818793)

**There are several other methods**

- You can train your model for a size like 256X256 and whenever you receive a image like 1000X1000 you can divide it into four images of size 250X250 and pad them with zeroes to make 4 images of 256X256 then predict for all 4 of them and later combine them and resize to 1000X1000. This method works but it is not very good.   
- Have a look at [this](https://stats.stackexchange.com/questions/388859/is-it-possible-to-give-variable-sized-images-as-input-to-a-convolutional-neural) thread.

Thanks!

**Comment by Rstln** [comment]
Thanks for your advices! I have one more question, the loss function in the paper is the combination of cross entrophy and MSE, but in your model, the loss function was input only as MSE. Is there any methods to define the loss function?

**Comment by tushar-jain01** [comment]
> Thanks for your advices! I have one more question, the loss function in the paper is the combination of cross entrophy and MSE, but in your model, the loss function was input only as MSE. Is there any methods to define the loss function?

In the paper they are doing two tasks 1) Colorization 2) Image Classification so they have used losses of both types MSE and Cross entropy where cross-entropy is a classification loss. 

yes you can define your own loss function as follows
```python
class MyLoss(tf.keras.losses.Loss):
    self.mse = tf.keras.losses.MeanSquaredError()
    self.cross_entropy = tf.keras.losses.CategoricalCrossentropy()
    def call(self,y_true,y_pred):
        yt_img , yt_class = y_true
        ypred_img , ypred_class = y_pred
        mse_loss = self.mse(yt_img,ypred_img)
        ce_loss = self.cross_entropy(yt_class,ypred_class)
        alpha = 0.2
    return tf.reduce_sum(mse_loss+alpha*ce_loss)
```

**Comment by Rstln** [comment]
Thanks. I want to ask if the training set images are the same type(flowers for example), does global level network make any difference? There are not classification, so what information could global level get?
